### PR TITLE
[LZY-84] Auto choose pool using partial requirements

### PR DESCRIPTION
**Pull request by vhaldemar**
TBD:
1) Deal with current tests and make new ones
2) Examples
3) Docs

Also, I wouldn't say I like the current mess with Any/None/value/canonize and wanna do some understandable pipeline here.

### CODE REVIEW

File: pylzy/lzy/api/v1/__init__.py [hunk 1]
```diff
@@ -235,43 +247,49 @@ def workflow(
...
         local_modules_path: Optional[Sequence[str]] = None,
-        provisioning: Provisioning = Provisioning.default(),
         interactive: bool = True,
-        cpu_type: Optional[str] = None,
-        cpu_count: Optional[int] = None,
-        gpu_type: Optional[str] = None,
-        gpu_count: Optional[int] = None,
-        ram_size_gb: Optional[int] = None,
+        provisioning: Optional[Provisioning] = None,
+        cpu_type: Optional[StringRequirement] = None,
```
Comment (mrMakaronka) [inline pylzy/lzy/api/v1/__init__.py:252]: Why is default value `None` instead of `Any`?

Comment (vhaldemar) [reply pylzy/lzy/api/v1/__init__.py]: Because you can do:
```
with workflow(cpu_count=1):
    op(cpu_count=Any)
```

And from the other hand
```
with workflow(cpu_count=1):
    op()
    # or
    op(cpu_count=None)
```
should not override workflow's cpu_count value.

File: pylzy/lzy/api/v1/provisioning.py [hunk 2]
```diff
@@ -16,46 +76,178 @@ class GpuType(Enum):
...
+    def _filter_spec(self, spec: "Provisioning") -> bool:
+        if (
+            spec.cpu_count_final < self.cpu_count_final or
+            spec.gpu_count_final < self.gpu_count_final or
+            spec.ram_size_gb_final < self.ram_size_gb_final
+        ):
+            return False
+
+        if (
+            self.cpu_type_final and self.cpu_type_final != spec.cpu_type_final or
```
Comment (ArtoLord) [inline pylzy/lzy/api/v1/provisioning.py:172]: I don't like checking for empty string here.
Maybe check if type is Any?

File: pylzy/lzy/api/v1/provisioning.py [hunk 3]
```diff
@@ -16,46 +76,178 @@ class GpuType(Enum):
...
-    @staticmethod
-    def default() -> "Provisioning":
-        return Provisioning(
-            cpu_type=CpuType.CASCADE_LAKE.value,
-            cpu_count=2,
-            gpu_type=GpuType.NO_GPU.value,
-            gpu_count=0,
-            ram_size_gb=2,
-        )
+    cpu_type: StringRequirement = None
```
Comment (ArtoLord) [inline pylzy/lzy/api/v1/provisioning.py:106]: Why default is None here.
Mb Any?

File: pylzy/lzy/api/v1/provisioning.py [hunk 4]
```diff
@@ -1,6 +1,66 @@
-from dataclasses import dataclass
+import dataclasses
+
 from enum import Enum
-from typing import Optional
+from typing import (
+    Any as typingAny,
```
Comment (ArtoLord) [inline pylzy/lzy/api/v1/provisioning.py:5]: nit: Mb not rename Any, just use it as typing.Any?

File: pylzy/lzy/api/v1/call.py [hunk 5]
```diff
@@ -159,7 +159,6 @@ def lazy(*args, **kwargs):
             return f(*args, **kwargs)
 
         prov = active_workflow.provisioning.override(provisioning)
-        prov.validate()
```
Comment (ArtoLord) [inline pylzy/lzy/api/v1/call.py:162]: Why removing validation here?

**Review by ArtoLord — CHANGES_REQUESTED** [review]

File: pylzy/lzy/api/v1/provisioning.py (hunk 3 above)
Comment (vhaldemar) [reply pylzy/lzy/api/v1/provisioning.py:106]: I had this discussion earlier with @mrMakaronka at this PR.
tldr None is a synonym for "user does not pass anything" and `Any` - is an explicit value and we need to maintain this difference for a transparent merging of values.

File: pylzy/lzy/api/v1/provisioning.py (hunk 4 above)
Comment (vhaldemar) [reply pylzy/lzy/api/v1/provisioning.py:5]: I thought about it, but it will complicate the code, because I'm using my Any as a type, but also as a value. typing.Any, from the other hand - can't be instantiated :(
But I left a possibility for the user to use typing.Any here, just they's mypy tests will go crazy/

File: pylzy/lzy/api/v1/provisioning.py (hunk 3 above)
Comment (ArtoLord) [reply pylzy/lzy/api/v1/provisioning.py:106]: Ok

File: pylzy/lzy/api/v1/call.py (hunk 5 above)
Comment (vhaldemar) [reply pylzy/lzy/api/v1/call.py:162]: With the new logic of merging and pool inference, I think there is only one place with validation that makes sense - right before the inference itself, because there are maybe any non-valid combinations in the middle.
Also, new validation is much less intense.

File: pylzy/lzy/api/v1/provisioning.py (hunk 2 above)
Comment (ArtoLord) [reply pylzy/lzy/api/v1/provisioning.py:172]: Maybe return Union[str, AnyRequirement] from cpu_type_final, not just str?
It would be more obvious than just returning empty string as Any
